pkg/netlink/netlink_watcher.go

```diff
@@ -0,0 +1,1988 @@
+		FBDB  map[FDBKey]FdbEntryStruct
+		L2NDB map[L2NexthopKey]L2NexthopStruct
+	}
+	latestdb := NlDBCopy{RDB: LatestRoutes, NDB: LatestNexthop, FBDB: LatestFDB, L2NDB: LatestL2Nexthop}
+	olddb := NlDBCopy{RDB: Routes, NDB: Nexthops, FBDB: FDB, L2NDB: L2Nexthops}
+	var eventStr []interface{}
+	eventStr = append(eventStr, routeEventStr)
+	eventStr = append(eventStr, nexthopEventStr)
+	eventStr = append(eventStr, fdbEventStr)
+	eventStr = append(eventStr, l2nexthopEventStr)
+	// Routes
+	oldgenmap = make(map[interface{}]interface{})
+	latestgenmap = make(map[interface{}]interface{})
+	for k, v := range latestdb.RDB {
+		latestgenmap[k] = v
+	}
+	for k, v := range olddb.RDB {
+		oldgenmap[k] = v
+	}
+	notify_changes(latestgenmap, oldgenmap, eventStr[0].([]string))
+	// Nexthops
+	oldgenmap = make(map[interface{}]interface{})
+	latestgenmap = make(map[interface{}]interface{})
+	for k, v := range latestdb.NDB {
+		latestgenmap[k] = v
+	}
+	for k, v := range olddb.NDB {
+		oldgenmap[k] = v
+	}
+	notify_changes(latestgenmap, oldgenmap, eventStr[1].([]string))
+	// FDB
+	oldgenmap = make(map[interface{}]interface{})
+	latestgenmap = make(map[interface{}]interface{})
+	for k, v := range latestdb.FBDB {
+		latestgenmap[k] = v
+	}
+	for k, v := range olddb.FBDB {
+		oldgenmap[k] = v
+	}
+	notify_changes(latestgenmap, oldgenmap, eventStr[2].([]string))
+	// L2Nexthop
+	oldgenmap = make(map[interface{}]interface{})
+	latestgenmap = make(map[interface{}]interface{})
+	for k, v := range latestdb.L2NDB {
+		latestgenmap[k] = v
+	}
+	for k, v := range olddb.L2NDB {
+		oldgenmap[k] = v
+	}
+	notify_changes(latestgenmap, oldgenmap, eventStr[3].([]string))
+}
+
+// resyncWithKernel fun resyncs with kernal db
+func resyncWithKernel() {
+	// Build a new DB snapshot from netlink and other sources
+	readLatestNetlinkState()
+	// Annotate the latest DB entries
+	annotateDBEntries()
+	// Filter the latest DB to retain only entries to be installed
+	applyInstallFilters()
+	// Compute changes between current and latest DB versions and inform subscribers about the changes
+	notifyDBChanges()
+	Routes = LatestRoutes
+	Nexthops = LatestNexthop
+	Neighbors = LatestNeighbors
+	FDB = LatestFDB
+	L2Nexthops = LatestL2Nexthop
+	DeleteLatestDB()
+}
+
+// DeleteLatestDB deletes the latest db snap
+func DeleteLatestDB() {
+	LatestRoutes = make(map[RouteKey]RouteStruct)
+	LatestNeighbors = make(map[NeighKey]NeighStruct)
+	LatestNexthop = make(map[NexthopKey]NexthopStruct)
+	LatestFDB = make(map[FDBKey]FdbEntryStruct)
+	LatestL2Nexthop = make(map[L2NexthopKey]L2NexthopStruct)
+}
+
+// monitorNetlink moniters the netlink
+func monitorNetlink(_ bool) {
+	for !stopMonitoring {
+		log.Printf("netlink: Polling netlink databases.")
+		resyncWithKernel()
+		log.Printf("netlink: Polling netlink databases completed.")
+		time.Sleep(time.Duration(pollInterval) * time.Second)
+	}
+	log.Printf("netlink: Stopped periodic polling. Waiting for Infra DB cleanup to finish")
+	time.Sleep(2 * time.Second)
```

Comment:
we have tested many times, we haven't seen any issue with current configuration.

Reply:
why infradb cannot send a signal when it is done?

Reply:
Hello @artek-koltun,

The reason that we do not send a signal even if we wanted is because it would not help. There is some configuration  that is injected in the system that is dynamic (learned routes for example through BGP). For FRR for example takes some time to delete that dynammic configuration even if the FRR static configuration has been deleted. So we need to wait until the dust has been settled let's say and then we do a final resync in order to cleanup the pipeline correctly.

Reply:
can we poll and check if that was really deleted?

Reply:
No we cannot poll and check as we do not poll for specific routes for example because we do not know which routes has been advertised by FRR this is completely dynamic and out of our control. Instead we compare states. We poll for the latest states of routes for example and then we compare that to the old state of the routes in the system. If netlink watcher understands that there are some differences on the old and the new state then it will send a signal to the vendor plugin module to delete the old routes from the pipeline.

Reply:
Won't `vtysh -c "show bgp ipv4 unicast neighbors routes"` or `vtysh -c "show bgp ipv4 unicast neighbors <bgp-nbr-ip> routes"` in case of FRR help us here?

Reply:
Hello @artek-koltun,

The netlink watcher translates a filtered and annotated subset of the Linux Netlink routing, neighbor and L2 FDB state into an internal snapshot and sends the delta between two successive Netlink snapshots as add/modify/delete notifications to the vendor plugin for offload rule programming. It is not (and should not need to be) aware of how the Linux Netlink databases are populated/cleaned up by control planes including FRR and and Linux internal ARP/ND or MAC learning.

Even when all configuration is removed from FRR and automatically created "representor" interfaces are deleted from Linux, it can take small finite time until all dependent Netlink routing, neighbor and FDB entries are actually gone as well. To ensure that all associated dynamic forwarding state in the xPU data plane is indeed gone before we stop the netlink watcher component of the EVPN bridge, we have found it good enough to wait for 1-2 seconds for "the dust to settle" before we take the final snapshot for the cleanup.

In a production environment I consider a graceful shutdown of the EVPN bridge with cleanup of the xPU data plane rather unlikely to be happen. This is rather a feature for rapid testing of code updates during development. For production use, a robust and seamless recovery after various failures, with and without data plane loss, is more relevant, but not yet fully covered in the current PR.

Reply:
ack